### PR TITLE
[Utility] Add Process class

### PR DESCRIPTION
**Pull request by aciidb0mb3r**
This class aims to be replacement to current system() and popen() methods and a wrapper around posix_spawn

### CODE REVIEW

**Comment by aciidb0mb3r** [comment]
Need to add specific errors in Error enum

File: Sources/Utility/Process.swift [hunk 1]
```diff
@@ -0,0 +1,253 @@
...
+        posix_spawnattr_setflags(&attributes, Int16(flags))
+
+        #if os(macOS)
+            var fileActions: posix_spawn_file_actions_t? = nil
+        #else
+            var fileActions = posix_spawn_file_actions_t()
+        #endif
+        posix_spawn_file_actions_init(&fileActions)
+
+        // Workaround for https://sourceware.org/git/gitweb.cgi?p=glibc.git;h=89e435f3559c53084498e9baad22172b64429362
```
Comment (neonichu) [inline Sources/Utility/Process.swift]: If this is a glibc specific workaround, we should `#if os(Linux)` it, shouldn't we?

Comment (aciidb0mb3r) [inline Sources/Utility/Process.swift]: hmm maybe, I'll dig up the PR where it was originally added.

File: Sources/Utility/Process.swift [hunk 2]
```diff
@@ -0,0 +1,253 @@
...
+        
+        var output = ""
+        loop: while true {
+            let n = read(saveOutputPipe[0], &buf, N)
+            switch n {
+            case  -1:
+                if errno == EINTR {
+                    continue  // try again!
+                } else {
+                    print("read output err1")
```
Comment (aciidb0mb3r) [inline Sources/Utility/Process.swift]: ya this is temporary, I need to add all error cases

File: Sources/Utility/Process.swift [hunk 3]
```diff
@@ -0,0 +1,253 @@
...
+            }
+        }
+        self.output = output
+        // Close the read end of the output pipe.
+        close(saveOutputPipe[0])
+    }
+
+    public func interrupt() {
+        assert(launched, "The process is not yet launched.")
+        _ = libc.kill(processID, SIGINT)
```
Comment (neonichu) [inline Sources/Utility/Process.swift]: This doesn't send the signal to the whole process group. We need to either use `-processID` or `killpg` explicitly.

File: Sources/Utility/Process.swift [hunk 4]
```diff
@@ -0,0 +1,253 @@
...
+    }
+
+    public func kill() {
+        assert(launched, "The process is not yet launched.")
+        _ = libc.kill(processID, SIGKILL)
```
Comment (neonichu) [inline Sources/Utility/Process.swift]: This doesn't send the signal to the whole process group. We need to either use `-processID` or `killpg` explicitly.

File: Sources/Utility/Process.swift [hunk 5]
```diff
@@ -0,0 +1,253 @@
...
+    }
+}
+
+extension Process {
+    /// Execute a subprocess and block until it finishes execution.
+    ///
+    /// - Parameters:
+    ///   - arguments: The arugments for the subprocess.
```
Comment (neonichu) [inline Sources/Utility/Process.swift]: typo: `arguments`

File: Sources/Utility/Process.swift [hunk 6]
```diff
@@ -0,0 +1,253 @@
...
+    ///   - environment: The environment to pass to subprocess. By default the current process environment
+    ///     will be inherited.
+    ///   - includeStandardError: Include the subprocess's standard error output, otherwise it is printed on standard stream.
```
Comment (neonichu) [inline Sources/Utility/Process.swift]: I think it isn't entirely clear what "standard stream" means here -- do you mean we will be capturing `stderr` if this is `true` and otherwise just let it print?

Comment (aciidb0mb3r) [inline Sources/Utility/Process.swift]: Yeah but ignore this, I went back on this and always capture stderr— a la llbuild

File: Sources/Utility/Process.swift (hunk 1 above)
Comment (aciidb0mb3r) [inline Sources/Utility/Process.swift]: Ah here it is: https://github.com/apple/swift-package-manager/pull/680
Doesn't look like we need to `#if`

File: Sources/Utility/Process.swift [hunk 7]
```diff
@@ -0,0 +1,295 @@
...
+    public private(set) var processID = ProcessID()
+
+    /// The stream containing output of the process, data is available only after execution.
+    public private(set) var outputStream: OutputByteStream? = nil
+
+    /// If the subprocess has launched.
+    public private(set) var launched = false
+
+    /// If the subprocess received a signal.
+    public private(set) var signalled = false
```
Comment (neonichu) [inline Sources/Utility/Process.swift]: Are we planning to utilise this in client code?

**Comment by neonichu** [comment]
LGTM

File: Sources/Utility/Process.swift (hunk 7 above)
Comment (aciidb0mb3r) [inline Sources/Utility/Process.swift]: not really, i'll make it internal

Comment (neonichu) [inline Sources/Utility/Process.swift]: It doesn't seem to really be used internally, either (except for setting it to `true`).

Comment (aciidb0mb3r) [inline Sources/Utility/Process.swift]: Yeah I am not sure what to do, I am not entirely sure if we should throw or just record that a signal was caught

Comment (ddunbar) [inline Sources/Utility/Process.swift]: IMHO we should record this in a struct containing also the result code -- information available once the process has been terminated (that could also include the other information which `wait4` can return, like resource utilization).

I would include the output bytes in that struct, until we support the client receiving the stream directly.

File: Sources/Utility/Process.swift [hunk 8]
```diff
@@ -0,0 +1,295 @@
...
+
+    /// The pipes to redirect and save the output of the subprocess.
+    private var saveOutputPipe: [Int32] = [0, 0]
```
Comment (ddunbar) [inline Sources/Utility/Process.swift]: Usually you only want to keep one of these FDs (the one you area allowed to used). I also think Foundation's name here makes the correct usage more obvious.

File: Sources/Utility/Process.swift [hunk 9]
```diff
@@ -0,0 +1,295 @@
...
+    public let environment: [String: String]
+
+    /// The return code of the process, available once process has finished execution.
+    public private(set) var returnCode: Int32 = 0
+
+    /// The process id of the spawned process, available after the process is launched.
+    public private(set) var processID = ProcessID()
+
+    /// The stream containing output of the process, data is available only after execution.
+    public private(set) var outputStream: OutputByteStream? = nil
```
Comment (ddunbar) [inline Sources/Utility/Process.swift]: Technically this is _The stream which output of the process is written to_

File: Sources/Utility/Process.swift [hunk 10]
```diff
@@ -0,0 +1,295 @@
...
+        } else {
+            posix_spawn_file_actions_adddup2(&fileActions, 1, 1)
+            posix_spawn_file_actions_adddup2(&fileActions, 2, 2)
+        }
+
+        let argv: [UnsafeMutablePointer<CChar>?] = arguments.map{ $0.withCString(strdup) }
+        defer { for case let arg? in argv { free(arg) } }
+        
+        let env: [UnsafeMutablePointer<CChar>?] = environment.map{ "\($0.0)=\($0.1)".withCString(strdup) }
+        defer { for case let arg? in env { free(arg) } }
```
Comment (ddunbar) [inline Sources/Utility/Process.swift]: `var argv` + argv.append(nil) would be mildly more efficient than our `argv + [nil]` (used later)

File: Sources/Utility/Process.swift [hunk 11]
```diff
@@ -0,0 +1,295 @@
...
+        guard rv == 0 else {
+            throw SystemError.posix_spawn(rv, arguments)
+        }
+    }
+
+    /// Blocks the calling process until the subprocess finishes execution.
+    ///
+    /// - Throws: Process.Error
+    public func waitUntilExit() throws {
+        assert(launched, "The process is not yet launched.")
```
Comment (ddunbar) [inline Sources/Utility/Process.swift]: `precondition` is probably more appropriate.

File: Sources/Utility/Process.swift [hunk 12]
```diff
@@ -0,0 +1,295 @@
...
+        // FIXME: We should probably read the output from somewhere else, maybe in a seperate thread?
```
Comment (ddunbar) [inline Sources/Utility/Process.swift]: Yes, this has to be done.

As currently written the launched process may end up blocked waiting for its output pipe to be read from, so it is critical that we be consuming the output in the background.

File: Sources/Utility/Process.swift [hunk 13]
```diff
@@ -0,0 +1,295 @@
...
+        var exitStatus: Int32 = 0
+        var result = waitpid(processID, &exitStatus, 0)
+        while (result == -1 && errno == EINTR) {
+            result = waitpid(processID, &exitStatus, 0)
+        }
+        if result == -1 {
+            throw SystemError.waitpid(errno)
+        }
+
+        if exitStatus & 0x7f == 0 {
```
Comment (ddunbar) [inline Sources/Utility/Process.swift]: Let's get private helper functions for this, to make it explicit what they are doing, versus inlining the constants.

File: Sources/Utility/Process.swift [hunk 14]
```diff
@@ -0,0 +1,295 @@
...
+            self.returnCode = (exitStatus >> 8) & 0xff
+        } else {
+            // We received a signal.
+            // FIXME: We may need a better way to represent this.
+            signalled = true
```
Comment (ddunbar) [inline Sources/Utility/Process.swift]: We should record the exact signal number for clients to access.

File: Sources/Utility/Process.swift [hunk 15]
```diff
@@ -0,0 +1,295 @@
...
+        }
+    }
+
+    /// Reads the output from save output pipes.
+    ///
+    /// - Throws: Process.Error
+    private func readOutput(on stream: OutputByteStream) throws {
+        assert(saveOutputPipe != [0, 0], "Pipes not initialized.")
+        // Close the write end of the output pipe.
+        let rv = close(saveOutputPipe[1])
```
Comment (ddunbar) [inline Sources/Utility/Process.swift]: This close should be done immediately after spawning the process, and then discarded.

File: Sources/Utility/Process.swift [hunk 16]
```diff
@@ -0,0 +1,295 @@
...
+                if let str = String(validatingUTF8: buf) {
+                    stream <<< str
+                } else {
+                    throw SystemError.popen(EILSEQ, arguments[0])
+                }
+            }
+        }
+
+        // Close the read end of the output pipe.
+        close(saveOutputPipe[0])
```
Comment (ddunbar) [inline Sources/Utility/Process.swift]: We are creating this but never closing it if the client didn't provided an output stream.

File: Sources/Utility/Process.swift [hunk 17]
```diff
@@ -0,0 +1,295 @@
...
+    }
+
+    public func interrupt() {
```
Comment (ddunbar) [inline Sources/Utility/Process.swift]: Doc comment plz (in particular note this will signal all processes in the process group).

Comment (ddunbar) [inline Sources/Utility/Process.swift]: One function to just send any signal seems reasonable here, versus separate functions.

File: Sources/Utility/Process.swift [hunk 18]
```diff
@@ -0,0 +1,295 @@
...
+extension Process {
+    /// Execute a subprocess and block until it finishes execution.
+    ///
+    /// - Parameters:
+    ///   - arguments: The arguments for the subprocess.
+    ///   - environment: The environment to pass to subprocess. By default the current process environment
+    ///     will be inherited.
+    /// - Returns: The output of the subprocess
+    /// - Throws: Process.Error
+    static public func checkOutput(arguments: [String], environment: [String: String] = env()) throws -> String {
```
Comment (ddunbar) [inline Sources/Utility/Process.swift]: I appreciate the helper functions, but I kinda always hated these names that Python chose. What do they mean? Does it or does it not throw if the called process exits with an error? If it does, what about the output?

I suggest we start w/o these then debate them on their individual merits.

File: Sources/Utility/Process.swift (hunk 7 above)
Comment (aciidb0mb3r) [inline Sources/Utility/Process.swift]: I mean, we don't need to return the output stream or even keep it public

**Comment by sclukey** [comment]
So, not to be a bother or anything, but why is this needed? Swift Foundation [already has a `Process` class](https://developer.apple.com/reference/foundation/process) and I don't see any features that this one adds.

I know the Linux version of Foundation's `Process` isn't quite done (and it is also called `Task` for some reason), but I think it would be more helpful for this project and the whole community if that one were fixed/finished rather than adding a new one here.

**Comment by ddunbar** [comment]
@sclukey there are specific things we need in SwiftPM (like the ability to pass extra file descriptors to subprocesses) which Foundation doesn't support, so we ultimately need this even if Foundation on Linux was complete.

There are other ways to work around that particular problem, but they involve adding even more complexity, so this was our preferred choice even though, as you note, it is rather unfortunate.

**Comment by sclukey** [comment]
@ddunbar I see, well this version looks to be a proper subset of Foundation's Process, but I guess there's more coming later. Thanks

**Comment by aciidb0mb3r** [comment]
@ddunbar heavily updated

File: Sources/Utility/Process.swift [hunk 19]
```diff
@@ -0,0 +1,384 @@
...
+
+import enum POSIX.SystemError
+import libc
+import Basic
+
+/// The POSIX signals.
+// FIXME: Move this to POSIX layer.
+public enum Signal: Int32 {
+    /// Interrupt signal.
+    case interrupt
```
Comment (ddunbar) [inline Sources/Utility/Process.swift]: Can we use RawRepresentable here?

File: Sources/Utility/Process.swift [hunk 20]
```diff
@@ -0,0 +1,384 @@
...
+    }
+
+    /// The exit status of the process.
+    public let exitStatus: ExitStatus
+
+    /// The output of the process. Available only if the process was asked to redirect its output.
+    public let output: Result<String, AnyError>
+
+    /// Create an instance using the process exit code and output result.
+    fileprivate init(exitCode: Int32, output: Result<String, AnyError>) {
```
Comment (ddunbar) [inline Sources/Utility/Process.swift]: Technically, `exitCode` should be `exitStatus` here -- the code is the value returned from a non-signaled process.

File: Sources/Utility/Process.swift [hunk 21]
```diff
@@ -0,0 +1,384 @@
...
+        case terminated(code: Int32)
+
+        /// The process was terminated due to a signal.
+        case signalled(signal: Int32)
+    }
+
+    /// The exit status of the process.
+    public let exitStatus: ExitStatus
+
+    /// The output of the process. Available only if the process was asked to redirect its output.
```
Comment (ddunbar) [inline Sources/Utility/Process.swift]: Should this be optional, then?

File: Sources/Utility/Process.swift [hunk 22]
```diff
@@ -0,0 +1,384 @@
...
+/// Process allows spawning new subprocesses and working with them.
+public final class Process {
+
+    /// Typealias for process id type.
+    public typealias ProcessID = pid_t
+
+    /// The arguments to execute.
+    public let arguments: [String]
+
+    /// The enviornment with which the process was executed.
```
Comment (ddunbar) [inline Sources/Utility/Process.swift]: Sic `enviornment`

File: Sources/Utility/Process.swift [hunk 23]
```diff
@@ -0,0 +1,384 @@
...
+            }
+            // Open the write end of the pipe as stdout and stderr, if desired.
+            posix_spawn_file_actions_adddup2(&fileActions, outputPipe[1], 1)
+            posix_spawn_file_actions_adddup2(&fileActions, outputPipe[1], 2)
+
+            // Close the other ends of the pipe.
+            posix_spawn_file_actions_addclose(&fileActions, outputPipe[0])
+            posix_spawn_file_actions_addclose(&fileActions, outputPipe[1])
+        } else {
+            posix_spawn_file_actions_adddup2(&fileActions, 1, 1)
```
Comment (ddunbar) [inline Sources/Utility/Process.swift]: This is redundant, we can just eliminate it?

File: Sources/Utility/Process.swift [hunk 24]
```diff
@@ -0,0 +1,384 @@
...
+            posix_spawn_file_actions_adddup2(&fileActions, 2, 2)
+        }
+
+        var argv: [UnsafeMutablePointer<CChar>?] = arguments.map{ $0.withCString(strdup) }
+        argv.append(nil)
+        defer { for case let arg? in argv { free(arg) } }
```
Comment (ddunbar) [inline Sources/Utility/Process.swift]: I would put this loop before the append and `!` it.

File: Sources/Utility/Process.swift [hunk 25]
```diff
@@ -0,0 +1,384 @@
...
+        if WIFSIGNALED(exitCode) {
+            self.exitStatus = .signalled(signal: WTERMSIG(exitCode))
+        } else {
+            precondition(WIFEXITED(exitCode), "unexpected exit status \(exitCode)")
+            self.exitStatus = .terminated(code: WEXITSTATUS(exitCode))
+        }
+    }
+}
+
+/// Process allows spawning new subprocesses and working with them.
```
Comment (ddunbar) [inline Sources/Utility/Process.swift]: Let's document that this class isn't designed to be thread safe.

File: Sources/Utility/Process.swift [hunk 26]
```diff
@@ -0,0 +1,384 @@
...
+        readOutputError = error
+        // Close the read end of the output pipe.
+        close(fd)
+    }
+
+    /// Send a signal to the process.
+    ///
+    /// Note: This will signal all processes in the process group.
+    public func signal(_ signal: Signal) {
+        assert(launched, "The process is not yet launched.")
```
Comment (ddunbar) [inline Sources/Utility/Process.swift]: `&& result == nil`

File: Tests/UtilityTests/ProcessTests.swift [hunk 27]
```diff
@@ -0,0 +1,92 @@
...
+import Utility
+
+class ProcessTests: XCTestCase {
+
+    func testBasics() throws {
+        let process = Process(args: "echo", "hello")
+        try process.launch()
+        let result = try process.waitUntilExit()
+        XCTAssertEqual(try result.output.dematerialize(), "hello\n")
+        XCTAssertEqual(result.exitStatus, .terminated(code: 0))
```
Comment (ddunbar) [inline Tests/UtilityTests/ProcessTests.swift]: Let's also test a call to `false` here.

File: Tests/UtilityTests/ProcessTests.swift [hunk 28]
```diff
@@ -0,0 +1,92 @@
...
+    }
+
+    func testPopen() throws {
+        XCTAssertEqual(try Process.popen(arguments: ["echo", "hello"]).output.dematerialize(), "hello\n")
+
+        let path = AbsolutePath(#file).parentDirectory.appending(component: "ArgumentParserTests.swift")
+        XCTAssertGreaterThan(try Process.popen(args: "cat", path.asString).output.dematerialize().characters.count, 4096)
```
Comment (ddunbar) [inline Tests/UtilityTests/ProcessTests.swift]: I would prefer this test just wrote a temp file of N bytes and then `cat`d it, rather than rely on an unrelated file.

File: Tests/UtilityTests/ProcessTests.swift [hunk 29]
```diff
@@ -0,0 +1,92 @@
...
+
+        let result = try Process.popen(args: "cat", "/bin/cat")
```
Comment (ddunbar) [inline Tests/UtilityTests/ProcessTests.swift]: I'm not clear what this is testing... in general I think it would be good to have a few more comments describing what the test is trying to cover.

File: Tests/UtilityTests/ProcessTests.swift [hunk 30]
```diff
@@ -0,0 +1,92 @@
...
+        func script(_ name: String) -> String {
+            return AbsolutePath(#file).parentDirectory.appending(components: "processInputs", name).asString
+        }
+        func running(_ pid: String) throws -> Bool {
+            let p = try Process.popen(args: "ps", "-p", pid)
+            // This is a hack for when running just tests via docker.
+            if try p.output.dematerialize().contains("defunct") {
+                return false
+            }
+            return p.exitStatus == .terminated(code: 0)
```
Comment (ddunbar) [inline Tests/UtilityTests/ProcessTests.swift]: There is a simpler way to test if a process is running, use `getpgid`.

File: Tests/UtilityTests/ProcessTests.swift [hunk 31]
```diff
@@ -0,0 +1,92 @@
...
+        }
+
+        // Test sigint terminates the script.
+        do {
+            let process = Process(args: script("print-pid"))
+            try process.launch()
+            usleep(10000) // Need some sleep to allow the process to launch.
```
Comment (ddunbar) [inline Tests/UtilityTests/ProcessTests.swift]: Why? This code shouldn't be able to race... if so I think there is a problem. The `process.launch` shouldn't be able to return before a PID is assigned, and I would hope that at any point once the test script is launched a SIGINT will cause it to end.

File: Sources/Utility/Process.swift [hunk 32]
```diff
@@ -0,0 +1,384 @@
...
+}
+
+/// Process result data which is available after process termination.
+public struct ProcessResult {
+    public enum ExitStatus {
+        /// The process was terminated normally with a exit code.
+        case terminated(code: Int32)
+
+        /// The process was terminated due to a signal.
+        case signalled(signal: Int32)
```
Comment (ddunbar) [inline Sources/Utility/Process.swift]: I don't think it is worth wrapping `Signal` if we don't also then go and use that wrapper type everywhere we store a signal.

File: Tests/UtilityTests/processInputs/print-pid [hunk 33]
```diff
@@ -0,0 +1,4 @@
+#!/bin/sh
+
+echo $$
+yes > /dev/null
```
Comment (ddunbar) [inline Tests/UtilityTests/processInputs/print-pid]: Given this test isn't using `yes`, it would be safer to write: `while true; do sleep 1; done`

This script should also start with: `set -e` to ensure it exits immediately whenever any of the subtasks has a problem. Without this, there is a race if the `echo` gets the SIGINT (which may be why you needed the sleep).

File: Tests/UtilityTests/ProcessTests.swift [hunk 34]
```diff
@@ -0,0 +1,92 @@
...
+            process.signal(.interrupt)
+            let result = try process.waitUntilExit()
+            XCTAssertEqual(result.exitStatus, .signalled(signal: Signal.interrupt.rawValue))
+            XCTAssertEqual("\(process.processID)\n", try result.output.dematerialize())
+            XCTAssertFalse(try running("\(process.processID)"))
+        }
+
+        // Test SIGKILL terminates the subprocess and any of its subprocess.
+        do {
+            let process = Process(args: script("subprocess"))
```
Comment (ddunbar) [inline Tests/UtilityTests/ProcessTests.swift]: I think this test would be easier to read if the subtask wrote a file containing its pid and its subprocess PID. You could just manually write a tidy `JSON` file using `printf(1)`

File: Sources/Utility/Process.swift (hunk 19 above)
Comment (aciidb0mb3r) [inline Sources/Utility/Process.swift]: No the signal constants are not imported as literals into swift

File: Tests/UtilityTests/ProcessTests.swift (hunk 31 above)
Comment (aciidb0mb3r) [inline Tests/UtilityTests/ProcessTests.swift]: I am not sure what could be the problem, the pid is returned but the process isn't able to do anything if its immediately sent a SIGINT

File: Sources/Utility/Process.swift (hunk 21 above)
Comment (aciidb0mb3r) [inline Sources/Utility/Process.swift]: idk, in practice, non-optional is easier to deal with even though it is kind of little incorrect

File: Sources/Utility/Process.swift (hunk 23 above)
Comment (aciidb0mb3r) [inline Sources/Utility/Process.swift]: This isn't redundant, without this it doesn't print on streams

File: Sources/Utility/Process.swift (hunk 24 above)
Comment (aciidb0mb3r) [inline Sources/Utility/Process.swift]: I don't think it matters where defer is?

File: Tests/UtilityTests/processInputs/print-pid (hunk 33 above)
Comment (aciidb0mb3r) [inline Tests/UtilityTests/processInputs/print-pid]: Actually the process not launching issue only happens on linux

File: Sources/Utility/Process.swift (hunk 24 above)
Comment (aciidb0mb3r) [inline Sources/Utility/Process.swift]: as in it captures the entire thing

Comment (ddunbar) [inline Sources/Utility/Process.swift]: Good point, thinko on my part

**Comment by aciidb0mb3r** [comment]
@swift-ci please test
